Release 11.7.0: An error occurred getting a credentia

**Expected Behaviour**
Run tests / pull images successfully (works in ~~11.6.0~~11.5.1)

**Actual Behaviour**
Setup of tests / containers fails
Running tests with DEBUG log enabled shows:

**Testcontainer Logs**
```
2025-10-01T07:03:36.216Z testcontainers [DEBUG] Pulling image "testcontainers/ryuk:0.14.0"...
2025-10-01T07:03:36.217Z testcontainers [DEBUG] Executing Docker credential provider "docker-credential-desktop"
2025-10-01T07:03:36.303Z testcontainers [ERROR] An error occurred getting a credential: 1
```

**Steps to Reproduce**
1. Install version 11.7.0
1. Run `await new MongoDBContainer("mongo:8").start()`
2. See error "An error occurred getting a credential"

**Environment Information**
- Operating System: OSX 15.7.1
- Docker Version: 
```
Client: Docker Engine - Community
 Version:           28.4.0
 API version:       1.51
 Go version:        go1.25.0
 Git commit:        d8eb465f86
 Built:             Wed Sep  3 20:52:20 2025
 OS/Arch:           darwin/arm64
 Context:           desktop-linux

Server: Docker Desktop 4.47.0 (206054)
 Engine:
  Version:          28.4.0
  API version:      1.51 (minimum version 1.24)
  Go version:       go1.24.7
  Git commit:       249d679
  Built:            Wed Sep  3 20:58:53 2025
  OS/Arch:          linux/arm64
  Experimental:     false
 containerd:
  Version:          1.7.27
  GitCommit:        05044ec0a9a75232cad458027ca83437aae3f4da
 runc:
  Version:          1.2.5
  GitCommit:        v1.2.5-0-g59923ef
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

- Node version: 24.9.0
- Testcontainers version: 11.7.0

--- a/packages/testcontainers/src/container-runtime/auth/credential-provider.test.ts
+++ b/packages/testcontainers/src/container-runtime/auth/credential-provider.test.ts
@@ -21,7 +21,7 @@ describe.sequential("CredentialProvider", () => {
   });
 
   it("should return the auth config for a registry", async () => {
-    mockSpawnReturns(
+    mockSpawnEmitsData(
       0,
       JSON.stringify({
         ServerURL: "registry",
@@ -40,7 +40,7 @@ describe.sequential("CredentialProvider", () => {
   });
 
   it("should default to the registry url when the server url is not returned", async () => {
-    mockSpawnReturns(
+    mockSpawnEmitsData(
       0,
       JSON.stringify({
         Username: "username",
@@ -61,8 +61,8 @@ describe.sequential("CredentialProvider", () => {
     expect(await credentialProvider.getAuthConfig("registry", containerRuntimeConfig)).toBeUndefined();
   });
 
-  it("should throw when get credentials fails", async () => {
-    mockSpawnReturns(
+  it("should return undefined when get credentials fails because we lookup optimistically", async () => {
+    mockSpawnEmitsData(
       1,
       JSON.stringify({
         ServerURL: "registry",
@@ -71,13 +71,19 @@ describe.sequential("CredentialProvider", () => {
       })
     );
 
+    expect(await credentialProvider.getAuthConfig("registry", containerRuntimeConfig)).toBeUndefined();
+  });
+
+  it("should throw when credential provider emits error", async () => {
+    mockSpawnEmitsError("ERROR");
+
     await expect(() => credentialProvider.getAuthConfig("registry", containerRuntimeConfig)).rejects.toThrow(
-      "An error occurred getting a credential"
+      "Error from Docker credential provider: Error: ERROR"
     );
   });
 
   it("should throw when get credentials output cannot be parsed", async () => {
-    mockSpawnReturns(0, "CANNOT_PARSE");
+    mockSpawnEmitsData(0, "CANNOT_PARSE");
 
     await expect(() => credentialProvider.getAuthConfig("registry", containerRuntimeConfig)).rejects.toThrow(
       "Unexpected response from Docker credential provider GET command"
@@ -85,7 +91,7 @@ describe.sequential("CredentialProvider", () => {
   });
 });
 
-function mockSpawnReturns(exitCode: number, stdout: string) {
+function mockSpawnEmitsData(exitCode: number, stdout: string) {
   const sink = new EventEmitter() as ChildProcess;
 
   sink.stdout = new Readable({
@@ -98,6 +104,21 @@ function mockSpawnReturns(exitCode: number, stdout: string) {
     write() {
       sink.stdout?.emit("data", stdout);
       sink.emit("close", exitCode);
+    },
+  });
+
+  mockSpawn.mockReturnValueOnce(sink);
+}
+
+function mockSpawnEmitsError(message: string) {
+  const sink = new EventEmitter() as ChildProcess;
+
+  sink.kill = () => true;
+  sink.stdout = new Readable({ read() {} });
+  sink.stdin = new Writable({
+    write(_chunk, _enc, cb) {
+      sink.emit("error", new Error(message));
+      cb?.();
     },
   });
 

--- a/packages/testcontainers/src/container-runtime/auth/credential-provider.ts
+++ b/packages/testcontainers/src/container-runtime/auth/credential-provider.ts
@@ -1,7 +1,7 @@
 import { spawn } from "child_process";
 import { log } from "../../common";
 import { RegistryAuthLocator } from "./registry-auth-locator";
-import { AuthConfig, ContainerRuntimeConfig, CredentialProviderGetResponse } from "./types";
+import { AuthConfig, ContainerRuntimeConfig } from "./types";
 
 export abstract class CredentialProvider implements RegistryAuthLocator {
   abstract getName(): string;
@@ -17,32 +17,35 @@ export abstract class CredentialProvider implements RegistryAuthLocator {
     const programName = `docker-credential-${credentialProviderName}`;
     log.debug(`Executing Docker credential provider "${programName}"`);
 
-    const response = await this.runCredentialProvider(registry, programName);
-
-    return {
-      username: response.Username,
-      password: response.Secret,
-      registryAddress: response.ServerURL ?? registry,
-    };
+    return await this.runCredentialProvider(registry, programName);
   }
 
-  private runCredentialProvider(registry: string, providerName: string): Promise<CredentialProviderGetResponse> {
+  private runCredentialProvider(registry: string, providerName: string): Promise<AuthConfig | undefined> {
     return new Promise((resolve, reject) => {
       const sink = spawn(providerName, ["get"]);
 
       const chunks: string[] = [];
       sink.stdout.on("data", (chunk) => chunks.push(chunk));
 
+      sink.on("error", (err) => {
+        log.error(`Error from Docker credential provider: ${err}`);
+        sink.kill("SIGKILL");
+        reject(new Error(`Error from Docker credential provider: ${err}`));
+      });
+
       sink.on("close", (code) => {
         if (code !== 0) {
-          log.error(`An error occurred getting a credential: ${code}`);
-          return reject(new Error("An error occurred getting a credential"));
+          return resolve(undefined);
         }
 
         const response = chunks.join("");
         try {
           const parsedResponse = JSON.parse(response);
-          return resolve(parsedResponse);
+          return resolve({
+            username: parsedResponse.Username,
+            password: parsedResponse.Secret,
+            registryAddress: parsedResponse.ServerURL ?? registry,
+          });
         } catch (e) {
           log.error(`Unexpected response from Docker credential provider GET command: "${response}"`);
           return reject(new Error("Unexpected response from Docker credential provider GET command"));

--- a/packages/testcontainers/src/container-runtime/auth/get-auth-config.ts
+++ b/packages/testcontainers/src/container-runtime/auth/get-auth-config.ts
@@ -55,7 +55,7 @@ export const getAuthConfig = async (registry: string): Promise<AuthConfig | unde
     }
   }
 
-  log.debug(`No registry auth locator found for registry "${registry}"`);
+  log.debug(`No auth config found for registry "${registry}"`);
   authsCache.set(registry, undefined);
   return undefined;
 };

